### DISCUSSION TITLE
Plotter's mode=1 strange behaviour

### DISCUSSION THREAD

**Discussion by sergei9838** [marcomusy/vedo#1045] [Q&A] [answered]
Hi, Marco,
This is what I have discovered in the development version of vedo.

mode=1 option in Plotter is supposed to be rotating the actor. Rotate the cube in the renderer somehow and save the camera position:
```
from vedo import *
C = Cube()
cam1 = show(C, mode=1).camera
close()

```
Get the camera position:
```
cam1.GetPosition()
# (0.0, 0.0, 3.3460652149512318)
```
The camera is in the default position, as expected for mode=1. The view of the cube is as expected too:

```
show(C).close()
```
<img width="475" alt="Screenshot 2024-02-08 at 22 15 39" src="https://github.com/marcomusy/vedo/assets/105316522/451c2074-809c-4219-b256-c29d46f61480">

BUT: the vertices of the cube are at the initial position!

```
C.vertices
# array([[-0.5, -0.5, -0.5],
#       [-0.5, -0.5,  0.5],
#       [-0.5,  0.5, -0.5],
...
```
So what figure does the plotter show then?! 
Moreover, let's clone the cube and display both:

```
C1 = C.clone()
show(C, C1.c("yellow"), axes=1).close()

```
<img width="414" alt="Screenshot 2024-02-08 at 22 25 37" src="https://github.com/marcomusy/vedo/assets/105316522/2e6e2fd4-0924-4fc5-9db4-8bb4a9be0be2">

So C1 is the correct copy of the initial cube. And not of the rotated C shown in the plotter.
This does not happen in vedo 2023.4.6: `C.points()` show rotated positions.
I am puzzled )
Cheers,
Sergei

**Comment by marcomusy** [comment]
Hi, that happens because you are changing the _representation_ of the object on the gpu but not the underlying object. You can still apply the transform from the actor with:
```py
from vedo import *
C = Cube()
cam1 = show(C, mode=1).camera
close()

C.apply_transform_from_actor()
print(C.vertices)
```

**Reply by sergei9838** [reply]
Thanks Marco, but now I am even more confused. If you call `show(C)` the second time, it will display the rotated cube from the first call. It means this GPU representation is stored somewhere in C. I guess in C.actor since `show(C.actor)` produces identical view to `show(C)`. But then it means that calling `show(C)` I am actually viewing `C.actor`? How then I can be sure calling `show` on a mesh that I am actually viewing the mesh and not its actor from some previous calls? 

This was not the way it worked in the earlier versions of vedo, as I mentioned. It would be more logical to `apply_transform_from_actor()` on the window's quit keeping the object and its actor in sync. This is my understanding of mode=1, otherwise it's identical to mode=0 plus additional confusion.

PS. Thus C1=C.clone() does not copy its actor since C1.apply_transform_from_actor() returns the identical transform. Is it intended?

Have a nice weekend, Marco, and thank you for your efforts and time!
Sergei

**Comment by marcomusy** [comment]
> This is my understanding of mode=1, otherwise it's identical to mode=0 plus additional confusion.

this (an others) are very good points.
There is no ambiguity with the actor because now vedo does not expose any api to modify the actor-related `vtkTransform` object.
so all actors _always_ have Identity transform.
Except in `.user_mode(1)` and  `.user_mode(3)`: for this modes we should indeed check if actors on screen were modified.

I'm flagging this as a bug to fix, thanks for forcing me to think better :)

**Comment by marcomusy** [comment]
You can try with the latest dev?

**Reply by sergei9838** [reply]
Perfect - thank you for the fast resolution! 
Have a great week
Sergei

**Reply by sergei9838** [reply]
Oh, just noticed that MousePan() mode also creates discrepancy between the mesh and its actor

**Reply by marcomusy** [reply]
Uhm.. this seems to work fine:
```py
from vedo import *

cube = Cube()
mode = interactor_modes.MousePan()

plt = Plotter(axes=1)
plt.user_mode(mode)
plt.show(cube)

print(cube.actor.GetMatrix())
```

I always get:
```
vtkMatrix4x4 (0x600001be0800)
  Debug: Off
  Modified Time: 194
  Reference Count: 2
  Registered Events: (none)
  Elements:
    1 0 0 0
    0 1 0 0
    0 0 1 0
    0 0 0 1
```

**Reply by sergei9838** [reply]
Probably the issue is not with MousePan but with viewing Image. 
```
dog = Image(dataurl+"images/dog.jpg")
show(dog.shift((100,100,0))).close()

```
produces error:
```
[vedo.plotter] INFO: object 'Image' was manually moved. Updated to its current position.
...
    a.retrieve_object().apply_transform_from_actor()
AttributeError: 'Image' object has no attribute 'apply_transform_from_actor'

```

**Reply by marcomusy** [reply, marked as answer]
try again on the latest dev..

**Reply by sergei9838** [reply]
Yes, no complaints any more!
One small suggestion: `scale` for Image does not have `origin=` option . Could you implement it? It is important for making a stencil from image standing in from of the object like I showed in [this](https://github.com/marcomusy/vedo/discussions/1001) discussion. 
Right now I am doing it like this:
```
S = LinearTransform()
S.scale(s=0.9, origin=camera_position)
dog.apply_transform(S)

```
Not much of overhead, but, I think, it would be more unified to have the origin option.

**Reply by marcomusy** [reply]
Makes sense! Would you mind opening a PR, you just need add `origin=True` in `visual.Actor3DHelper.scale()`.

**Reply by sergei9838** [reply]
Yes, this works just fine - thank you!!!
```
def scale(self, s, origin=True):
    """Multiply object size by `s` factor with the homothetic center in `origin`"""
    LT = vedo.LinearTransform().scale(s, origin)
    return self.apply_transform(LT)

```
